[Chong Chan How] iP

ChillinRage opened the pull request:
# SomeBOTy
> "Somebody once told me the world is gonna roll me." ~Shrek Probably.

This is some task bot to help you track tasks. 

- [x] Simple 
- [x] Easy to use
- [x] Text-based 

All you need to do is:
1. Download it [not available yet lol](https://youtu.be/dQw4w9WgXcQ).
2. Double-click to launch it.
3. Now you're set! 😉

**Features**
- Managing Tasks
- Managing details related to a task (coming soon)
- Reminders (coming not so soon)

Link to user guide can be found [maybe coming soon](https://youtu.be/dQw4w9WgXcQ).

If you are a java programmer, you can customize it for your own ~~fetish~~ taste too! Here's the main method:
```java
public class Launcher {
    public static void main(String[] args) {
        try {
            // starts the application at given class.
            Application.launch(Main.class, args);

        } catch (Exception e) {
            e.printStackTrace();
        }

    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ChillinRage committed:
Level-0

Implement Rename, Greet, Exit.

ChillinRage committed:
Level-1

1) renamed file back to Duke.
2) Added echo.

ChillinRage committed:
Level-2

Added Add and List.

ChillinRage committed:
Level-3

Added mark/unmark feature.

ChillinRage committed:
Level-4

Added 3 subclasses of "Task" and added more 3 respective commands to Duke.

ChillinRage committed:
Level-5

added test cases into input.txt and expected display results in EXPECTED.txt.

ChillinRage committed:
Level-5

Added error handling for the following:
- empty descriptions for all relevant commands
- invalid formatting for "event" and "deadline" commands.
- non-numerical and out-of-bounds index for mark/unmark commands.
- unknown commands

ChillinRage committed:
Level-6

Added "delete" function to remove a task from the list.

Added special message upon using "list" command on an empty list.

Switched the data type for storing the list of tasks to ArrayList.

ChillinRage committed:
Level-7

- Added fileManager.java

Handles saving and writing of task list to tasks.csv file.

- Updated Duke.java to use fileManager to fetch task list (if any) from data.csv on initialize, and overwrites the csv file with update list on exit.

- Added an abstract function in Task.java to convert the task details into a string line to store in csv file. ToDo, Deadline and Event will provide their own implementations.

ChillinRage committed:
Level-8

Updated Duke to recognize and parse date and time from input.

Added dateManager to handle parsing and output formatting of date and time.

Deadline and Event no longer holds their dates as a string, but as type "LocalDateTime".

Changed the structure of saving deadline and event types. Now it follows a fixed format of "dd MMM yyyy HH:mm".

ChillinRage committed:
A-MoreOOP

Updated Duke to be more OOP friendly.

Added taskManager to manage task related actions, exclusding fetching and storing of task data in csv file. That is still being maintained by fileManager.

Added commandManager to parse input and separate it into command word and description. It then directs the description to the appropriate managers and forwards the responses to Response class.

Added Response.java to handle the outputs of the Duke bot. In particular, it handles the reading of input and passing it to commandManager, then formats and outputs the responses appropriately to the interface.

ChillinRage committed:
A-Packages

moved all files in "main/java" into new folder "someboty" and is treated as a package.

Changed the name of the bot from "Duke" to "someboty" (I know, very creative).

ChillinRage committed:
Merge branch 'master' into add-gradle-support

ChillinRage committed:
Delete Duke.class

ChillinRage committed:
A-Gradle

Merged A-Gradle branch to main branch  and set up gradle in Intellij.

Fixed the file path to tasks.csv file in fileManager. Now the path file is relative to root of project directory instead of current directory.

ChillinRage committed:
A-JUnit

Added JUnit tests, specifically for "taskManager" and "dateManager".

"taskManagerTest" tests on correct parsing and formatting of newly created tasks, and ensuring that task list holds the correct number of tasks (i.e. use invalid index when deleting a task should not decrease or increase the task list).

"dateManagerTest" tests on correct parsing of valid and invalid date formats.

Improved OOP-ness of code, specifically moving the checking of task type and adding of new tasks from "taskManager" to "Task".

ChillinRage committed:
A-Jar

Packaged "someboty" app into a jar file and released version 0.1 (v0.1) on Github.

Changed the initial file path in "someBOTy" to be the current directory.

Added checking of existing directory "data" in current file path and create one if necessary.

ChillinRage committed:
Level-9

Added "find" feature to allow user to find tasks that matches the given substring.

Changed return type of "addTask", "deleteTask" and "setTaskStatus" in "taskManager" to Task (previously String). The formatting of output string will now be handled in commandManager.

Fixed bug that causes the fileWriter to throw an error when trying to write to file.

ChillinRage committed:
A-JavaDoc

Provided JavaDoc comments to most of the non-private (and some private) methods and classes.

ChillinRage committed:
A-CodingStandard

Tweaked some code to comply with the coding standard.

ruijietay commented on src/main/java/someboty/Managers/responseManager.java:
```diff
@@ -0,0 +1,72 @@
+     * Prompts user for input and forwards it to the command center.
+     * Then receives a response from it and prints it onto interface.
+     */
+    public void parseInput() {
```
Would this be better if the method name `parseInputs()` specifies what is being parsed?

ruijietay commented on src/main/java/someboty/Managers/taskManager.java:
```diff
@@ -0,0 +1,145 @@
+     * @return A list of tasks with names that contain the substring.
+     */
+    protected ArrayList<Task> findTasks(String substring) {
+        ArrayList<Task> matchList = new ArrayList<>();
```
Would this benefit from a clearer variable name? Without reading the Javadocs, I was unsure whether the list contains Task variables that needs to be matched, or whether it contains Task variables that matches a given string.

ruijietay commented on src/main/java/someboty/someBOTy.java:
```diff
@@ -0,0 +1,38 @@
+import someboty.Managers.taskManager;
+
+// main class for the project
+public class someBOTy {
```
Would this be better if the class started with a capital letter? May I also clarify if BOT is an abbreviation since it is all in capital letters?

ruijietay commented on src/main/java/someboty/Managers/commandManager.java:
```diff
@@ -0,0 +1,227 @@
+     * @return String response after deleting the task.
+     * @throws InputException Description is not an integer.
+     */
+    private String deleteTask(String description) throws InputException {
```
Good naming of method!

ruijietay commented on src/main/java/someboty/Managers/dateManager.java:
```diff
@@ -0,0 +1,77 @@
+ */
+public class DateManager {
+
+    private static final DateTimeFormatter OUT_FORMATT = DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm");
```
Would this benefit from a clearer variable name instead of `OUT_FORMATT`?

ruijietay reviewed (commented):
Overall, I like that you used nouns for classes/variables and verbs for methods/functions! The code is generally neat with proper naming whenever appropriate. Good job 😄

ChillinRage committed:
Added Launcher

RingoftheKing commented on src/main/java/someboty/Managers/dateManager.java:
```diff
@@ -0,0 +1,77 @@
+    private static final DateTimeFormatter OUT_FORMATT = DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm");
+
+    // accepted date formats
+    private static final String[] DATE_FORMATS = {
+        "yyyy-MM-dd HHmm",
+        "dd-MM-yyyy HHmm",
+        "MM-dd-yyyy HHmm",
+        "dd/MM/yyyy HHmm",
+        "MM/dd/yyyy HHmm",
+        "yyyy/MM/dd HHmm",
+        "dd MMM yyyy HH:mm"
+    };
```
Your method of making date formatting more robust is pretty neat. It allows code to be more resilient to various date formats so nice job

RingoftheKing reviewed (commented):
LGTM, your code is very nicely following coding standards and neccessary abstractions are well done

ChillinRage committed:
Added GUI

Added a GUI for the task bot.

ChillinRage committed:
A-Assertions

Added assertion statements to TaskManager to check if tasks are properly fetched and saved from file, and if a new task is successfully created.

ChillinRage committed:
Rename fileManager.java to FileManager.java

ChillinRage committed:
Rename commandManager.java to CommandManager.java

ChillinRage committed:
Rename commandManager.java to CommandManager.java

ChillinRage committed:
Rename dateManager.java to DateManager.java

ChillinRage committed:
Rename fileManager.java to FileManager.java

ChillinRage committed:
Rename taskManager.java to TaskManager.java

ChillinRage committed:
Improve code quality

Package name style: all package names are in lower case

Package names had its first character capitalized (e.g. Managers).

Package names are written in all lower case to avoid conflict with
the names of classes or interfaces.

Method header comments: improve formatting of header
comments

Some previous header comments were presented in a less
reader-friendly format.

ChillinRage committed:
Merge branch 'branch-A-CodeQuality'

ChillinRage committed:
Merge branch 'branch-a-assertions'

ChillinRage committed:
Add "save" command

Add new command: "save"

The only time someBOTy saves the task list is when the
user closes the program through the "bye" command.
With the implementation of a GUI application, users
might close the application through other means like
clicking on the close window button or using ALT+F4,
bypassing the task file-saving process.

Users can save their task list through a "save"
command.

This allows users to control what tasks they want to
be saved in the task file.

ChillinRage committed:
Update logic flow and response

MainWindow class: handle TerminateException

With the implementation of a GUI, MainWindow class
now handles the running and termination of the application.
As such, the "bye" command is no longer able to close the
application.

MainWindow catch and handle TerminateException.

This allows MainWindow to know if the application
needs to be terminated.

ChillinRage committed:
Rename folder GUI to gui

ChillinRage committed:
Rename Main.java to Main.java

ChillinRage committed:
Rename MainWindow.java to MainWindow.java

ChillinRage committed:
Rename folder Manager to manager

ChillinRage committed:
Rename folder Managers to managers

ChillinRage committed:
Rename FileManager.java to FileManager.java

ChillinRage committed:
Rename TaskManager.java to TaskManager.java

ChillinRage committed:
Rename folder Tasks to tasks

ChillinRage committed:
Rename Event.java to Event.java

ChillinRage committed:
Rename Task.java to Task.java

ChillinRage committed:
Rename ToDo.java to ToDo.java

ChillinRage committed:
Rename folder Exceptions to exceptions

ChillinRage committed:
Rename TerminateException.java to TerminateException.java

ChillinRage committed:
Rename someBOTy.java to SomeBoty.java

ChillinRage committed:
Add expense extension and some code improvements

Add expense feature to bot

A new class Expense is created to store information for every
expense added.

ExpenseManager is created to handle expenses created and stored,
similar to TaskManager.

Added more abstraction to fileManager,
- Each task class now handles the conversion of a csv line to
   its instance creation.
- A separate method to get scanner and fileWriter to allow
   reusability of code for fetching both task and expense files.

ChillinRage committed:
fixed the manager Testers

Renamed the managerTests to fit the coding standard.

ChillinRage committed:
Rename dateManagerTest.java to DateManagerTest.java

ChillinRage committed:
Rename taskManagerTest.java to TaskManagerTest.java

ChillinRage committed:
Update Expense.java

ChillinRage committed:
Merge pull request #2 from ChillinRage/branch-D-Expenses

Update Expense.java

ChillinRage committed:
Added application title to Main

Set a title for the window application.

ChillinRage committed:
Add files via upload

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs

ChillinRage committed:
Update README.md

Added a user guide to the bot.

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs

ChillinRage committed:
Added Ui.png

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs

ChillinRage committed:
Updated bot exit message

Changed the exit message that the bot prints.

ChillinRage committed:
Improved UI and responses

Changed some bot greeting format.

Changed 'toString()' method of task subtypes.

ChillinRage committed:
Delete docs/Ui.png

ChillinRage committed:
Added Ui.png to docs
